src/map/mod.rs

```diff
@@ -0,0 +1,315 @@
+        f(&mut MapChildElement)
+    }
+
+    fn skip(&mut self, _n: usize) {}
+
+    fn delete<R>(&mut self, f: impl FnOnce(Mut<'_, MapChildElement>) -> R) -> R {
+        f(&mut MapChildElement)
+    }
+}
+
+/// This is a marker trait that is used to pass only map-specific children
+/// to the [`map`] function.
+pub trait MapChild<State, Action>:
+    View<State, Action, ViewCtx, MapMessage, Element = MapChildElement>
+{
+}
+
+impl<V, State, Action> MapChild<State, Action> for V where
+    V: View<State, Action, ViewCtx, MapMessage, Element = MapChildElement>
+{
+}
+
+pub struct Map<State, Action, Children> {
+    map_view: MapDomView<State, Action>,
+    children: Children,
+    zoom: Option<f64>,
+    center: Option<(f64, f64)>,
+    phantom: PhantomData<fn() -> (State, Action)>,
+}
+
+impl<State, Action, Children> Map<State, Action, Children> {
+    pub fn zoom(mut self, value: f64) -> Self {
+        self.zoom = Some(value);
+        self
+    }
+
+    pub fn on_zoom_end<F>(self, callback: F) -> Map<State, Action, (Children, OnZoomEnd<State, F>)>
+    where
+        F: Fn(&mut State, f64) + 'static,
+    {
+        let Self {
+            map_view,
+            children,
+            zoom,
+            center,
+            phantom,
+        } = self;
+        let children = (children, on_zoom_end(callback));
+        Map {
+            map_view,
+            children,
+            zoom,
+            center,
+            phantom,
+        }
+    }
+
+    pub fn center(mut self, lat: f64, lng: f64) -> Self {
+        self.center = Some((lat, lng));
+        self
+    }
+}
+
+impl<State, Action, Children> ViewMarker for Map<State, Action, Children> {}
+
+pub struct MapViewState<DS, CS> {
+    map_dom_state: DS,
+    children_state: CS,
+    map_ctx: MapCtx,
+}
+
+#[derive(Debug, Clone)]
+pub enum MapMessage {
+    InitMap,
+    ZoomEnd(f64), // TODO: remove this
+}
+
+/// Distinctive ID for better debugging
+const MAP_VIEW_ID: ViewId = ViewId::new(1236068);
+
+impl<State, Action, Children> View<State, Action, ViewCtx, DynMessage>
+    for Map<State, Action, Children>
+where
+    State: 'static,
+    Action: 'static,
+    Children: MapChildren<State>,
+{
+    type Element = <MapDomView<State, Action> as View<State, Action, ViewCtx, DynMessage>>::Element;
+
+    type ViewState = MapViewState<
+        <MapDomView<State, Action> as View<State, Action, ViewCtx, DynMessage>>::ViewState,
+        Children::SeqState,
+    >;
+
+    fn build(&self, ctx: &mut ViewCtx) -> (Self::Element, Self::ViewState) {
+        ctx.with_id(MAP_VIEW_ID, |ctx| {
+            let (map_dom_element, map_dom_state) = self.map_view.build(ctx);
+
+            let map_options = leaflet::MapOptions::default();
+            let leaflet_map = leaflet::Map::new_with_element(&map_dom_element.node, &map_options);
+
+            let mut elements = AppendVec::default();
+            let mut map_ctx = MapCtx::new(leaflet_map.clone(), ctx.message_thunk());
```

Comment:
Hmm, I'm not sure it's a good idea to store that `MessageThunk` here... it's bound to exactly this view and not child views.

Actually thinking about it, I think the "easiest" is to bite the bullet (regarding named lifetimes) and just wrap the `ViewCtx` (from xilem_web`) as mutable reference inside the `MapCtx` and add expose this for the children. 

That would require creating a "temporary" `MapCtx` in rebuild.

Alternatively you need to copy `ctx.view_path()` into that id_path and provide your own `MessageThunk` (copy paste from xilem_web).

Reply:
> Actually thinking about it, I think the "easiest" is to bite the bullet (regarding named lifetimes) and just wrap the `ViewCtx` (from `xilem_web`) as mutable reference inside the `MapCtx` and add expose this for the children.

Yes, I tried that first. And I tried a lot... but somehow I'm too dumb for that :disappointed:

> Alternatively you need to copy `ctx.view_path()` into that id_path and provide your own `MessageThunk` (copy paste from xilem_web).

Right, that's what I had in mind afterwards, but I have no access to `AppRunner`, `AppMessage` etc.

Reply:
I just checked the lifetime route, and indeed, it's at the very least not trivial, as it needs to be in all the relevant type defs...
So in that case I'd just copy the path and to avoid allocs save the path/ctx in the state, as you do already.

> Right, that's what I had in mind afterwards, but I have no access to AppRunner, AppMessage etc.

Oh right, that's suboptimal, I guess we need to create an API/expose this, I'll think about this.